README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# mapbayr <img align="right" src = "inst/logo.png" width="135px">

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version-last-release/mapbayr)](https://CRAN.R-project.org/package=mapbayr)
[![](https://img.shields.io/badge/Citation-CPT:PSP-blue.svg)](https://pubmed.ncbi.nlm.nih.gov/34342170/)
<!-- badges: end -->

mapbayr is a free and open source package for *maximum a posteriori*
bayesian estimation of PK parameters in R. Thanks to a single function,
`mapbayest()`, you can estimate individual PK parameters from:

- a population PK model (coded in
  [mrgsolve](https://github.com/metrumresearchgroup/mrgsolve)),
- a data set with concentrations (NM-TRAN format).

It was designed to be easily wrapped in [shiny
apps](https://github.com/FelicienLL/mapbayr-shiny) in order to ease
model-based Therapeutic Drug Monitoring, also referred to as
Model-Informed Prediction Dosing (MIPD).

## Installation

mapbayr is available on
[CRAN](https://CRAN.R-project.org/package=mapbayr). You can install the
development version from github by executing the following code in R
console.

``` r
install.packages("devtools")
devtools::install_github("FelicienLL/mapbayr")
```

mapbayr relies on
[mrgsolve](https://github.com/metrumresearchgroup/mrgsolve) for model
implementation and ordinary differential equation solving which requires
C++ compilers. If you are a Windows user, you would probably need to
install [Rtools](https://cran.r-project.org/bin/windows/Rtools/). Please
refer to the [installation guide of
mrgsolve](https://github.com/metrumresearchgroup/mrgsolve/wiki/mrgsolve-Installation)
for additional information.

## Example

``` r
library(mapbayr)
library(mrgsolve)
```

#### 1) Properly code you model

``` r
code <- "
$PARAM @annotated
TVCL:  0.9 : Clearance
TVV1: 10.0 : Central volume
V2  : 10.0 : Peripheral volume of distribution
Q   :  1.0 : Intercompartmental clearance

ETA1: 0 : Clearance (L/h)
ETA2: 0 : Central volume (L)

$PARAM @annotated @covariates
BW : 70 : Body weight (kg)

$OMEGA 0.3 0.3
$SIGMA
0.05 // proportional
0.1 // additive

$CMT @annotated
CENT  : Central compartment (mg/L)[ADM, OBS]
PERIPH: Peripheral compartment ()

$TABLE
double DV = (CENT/V1) *(1 + EPS(1)) + EPS(2);

$MAIN
double CL = TVCL * exp(ETA1 + ETA(1)) * pow(BW / 70, 1.2) ;
double V1 = TVV1 * exp(ETA2 + ETA(2)) ;
double K12 = Q / V1  ;
double K21 = Q / V2  ;
double K10 = CL / V1 ;

$ODE
dxdt_CENT   =  K21 * PERIPH - (K10 + K12) * CENT ;
dxdt_PERIPH =  K12 * CENT - K21 * PERIPH ;

$CAPTURE DV CL
"

my_model <- mcode("Example_model", code)
```

#### 2) Bring your dataset

``` r
my_data <- data.frame(ID = 1, time = c(0,6,15,24), evid = c(1, rep(0,3)), cmt = 1, amt = c(100, rep(0,3)), 
                      rate = c(20, rep(0,3)), DV = c(NA, 3.9, 1.1, 2), mdv = c(1,0,0,1), BW = 90)
my_data
#>   ID time evid cmt amt rate  DV mdv BW
#> 1  1    0    1   1 100   20  NA   1 90
#> 2  1    6    0   1   0    0 3.9   0 90
#> 3  1   15    0   1   0    0 1.1   0 90
#> 4  1   24    0   1   0    0 2.0   1 90
```

#### 3) And estimate !

``` r
my_est <- mapbayest(my_model, data = my_data)
```

As building dataset into a NM-TRAN format can be painful, you can use
pipe-friendly `obs_rows()`, `adm_rows()` and `add_covariates()`
functions in order to pass administration and observation information,
and perform the estimation subsequently.

``` r
my_est <- my_model %>% 
  adm_rows(time = 0, amt = 100, rate = 20) %>% 
  obs_rows(time = 6, DV = 3.9) %>% 
  obs_rows(time = 20, DV = 1.1) %>% 
  obs_rows(time = 24, DV = 2, mdv = 1) %>% 
  add_covariates(BW = 90) %>% 
  mapbayest()
```

#### 4) Then, use the estimations

The results are returned in a single object (“mapbayests” S3 class)
which includes input (model and data), output (etas and tables) and
internal arguments passed to the internal algorithm (useful for
debugging). Additional methods are provided to ease visualization and
computation of a posteriori outcomes of interest.

``` r
print(my_est)
#> Model: Example_model 
#> ID : 1 individual(s).
#> OBS: 2 observation(s).
#> ETA: 2 parameter(s) to estimate.
#> 
#> Estimates: 
#>   ID      ETA1      ETA2
#> 1  1 0.3872104 0.1569604
#> 
#> Output (4 lines): 
#>   ID time evid cmt amt rate mdv  DV IPRED  PRED   CL BW  ETA1  ETA2
#> 1  1    0    1   1 100   20   1  NA 0.000 0.000 1.79 90 0.387 0.157
#> 2  1    6    0   1   0    0   0 3.9 4.162 5.174 1.79 90 0.387 0.157
#> 3  1   15    0   1   0    0   0 1.1 1.087 1.647 1.79 90 0.387 0.157
#> 4  1   24    0   1   0    0   1 2.0 0.556 0.959 1.79 90 0.387 0.157
```

``` r
plot(my_est)
```

![](man/figures/README-plot1-1.png)<!-- -->

``` r
hist(my_est)  
```

![](man/figures/README-plot2-1.png)<!-- -->

``` r
# Easily extract a posteriori parameter values to compute outcomes of interest
get_eta(my_est)
#>      ETA1      ETA2 
#> 0.3872104 0.1569604
get_param(my_est, "CL")
#> [1] 1.79217

# The `use_posterior()` functions updates the model object with posterior values and covariates to simulate like with a regular mrgsolve model
my_est %>% 
  use_posterior() %>% 
  data_set(expand.ev(amt = c(50, 100, 200, 500), dur = c(5, 24)) %>% mutate(rate = amt/dur)) %>% 
  carry_out(dur) %>% 
  mrgsim() %>% 
  plot(DV~time|factor(dur), scales = "same")
```

![](man/figures/README-plot3-1.png)<!-- -->

## Development

mapbayr is under development. Your feedback for additional feature
requests or bug reporting is welcome. Contact us through the [issue
tracker](https://github.com/FelicienLL/mapbayr/issues).

## Features

mapbayr is a generalization of the “MAP Bayes estimation” tutorial
available on the [mrgsolve
blog](https://mrgsolve.org/blog/posts/map_bayes.html). Additional
features are:

- a unique function to perform the estimation: `mapbayest()`.
- accepts a large variety of structural models thanks to the flexibility
  of mrgsolve
- flexibility with random effects on parameters, accepting both
  inter-individual and inter-occasion variability.
- additive, proportional, mixed or exponential (without prior
  log-transformation of data) residual error models.
- estimate from both parent drug and metabolite simultaneously.
- fit multiple patients stored in a single dataset.
- functions to easily pass administration and observation information,
  as well as plot methods to visualize predictions and parameter
  distribution.
- a single output object to ease post-processing, depending on the
  purpose of the estimation.
- several optimization algorithm available, such as “L-BFGS-B” (the
  default) or “newuoa”.
- handling data below the limit of quantification.
- estimate only a subset of ETAs defined in the model.
- flatten priors to favor observed data.

## Performance

Reliability of parameter estimation against NONMEM was assessed for a
wide variety of models and data. The results of this validation study
were published in [CPT:Pharmacometrics & System
Pharmacology](https://pubmed.ncbi.nlm.nih.gov/34342170/), and materials
are available in [a dedicated
repository](https://github.com/FelicienLL/mapbayr-CPTPSP-2021). If you
observe some discrepancies between mapbayr and NONMEM on your own model
and data, feel free to contact us through the [issue
tracker](https://github.com/FelicienLL/mapbayr/issues).

## *mrgsolve* model specification

mapbayr contains a library of example model files (.cpp), accessible
with `exmodel()`. You are invited to perform MAP-Bayesian estimation
with your own models. These model files should be slightly modified in
order to be “read” by mapbayr with the subsequent specifications:

### 1. `$PARAM` block

#### 1.1 ETA specifications

- Mandatory:
  - Add as many ETA as there are parameters to estimate (i.e. the length
    of the OMEGA matrix diagonal).
  - Name them as ETAn (n being the N° of ETA).
  - Set 0 as default value.
- Strongly recommended:
  - Provide a description as a plain text

``` c
$PARAM @annotated
ETA1 : 0 : CL (L/h)
ETA2 : 0 : VC (L)
ETA3 : 0 : F ()
//do not write ETA(1)
//do not write iETA
```

#### 1.2 Covariates

- Mandatory:
  - Use a `@covariates` tag to record covariates in the `$PARAM` block.
    Otherwise, you will not be allowed to pass a dataset with covariates
    columns.
  - Set the reference value.
- Strongly recommended
  - Provide a description as a plain text
  - Provide units in parentheses (or a description of 0/1 coding for
    categorical covariates)

``` c
$PARAM @annotated @covariates
BW : 70 : Body weight (kg)
SEX : 0 : Sex (0=Male, 1=Female)
```

### 2. `$CMT` block

- Strongly recommended…  
  … yet **mandatory** if you have multiple types of DV, i.e. parent
  drug + metabolite:
  - A `@annotated` tag must be used to record compartments.
  - Write OBS in brackets to define the observation compartment(s). Also
    used by `obs_rows()` to fill the ‘cmt’ column in your dataset.
  - Write ADM in brackets to define “default” administration
    compartment(s). This information is not used for optimization
    process and the `mapbayest()` function. The information is mandatory
    if you use `adm_rows()` to build your dataset in order to
    automatically set the value of the ‘cmt’ column. Especially useful
    if you use a model with an absorption from several depot compartment
    requiring to duplicate administrations lines in the data set.

``` c
//example: model with dual zero and first order absorption in compartment 1 & 2, respectively, and observation of parent drug + metabolite 
$CMT @annotated
DEPOT: Depot [ADM]
CENT_PAR: examplinib central [ADM, OBS]
PERIPH : examplinib peripheral
CENT_MET : methylexamplinib central [OBS] 
```

### 3. `$OMEGA` block

- Mandatory:
  - The length of the omega matrix must be the same as the number of
    ETAn provided in `$PARAM`.
  - The order of the omega values must correspond to the order of the
    ETAs provided in `$PARAM`. This cannot be checked by mapbayr !

``` c
$OMEGA
0.123 0.456 0.789
$OMEGA @block
0.111 
0.222 0.333
// reminder: omega values can be recorded in multiple $OMEGA blocks
```

### 4. `$SIGMA` block

The definition of the `$SIGMA` block may not be as straightforward as
other blocks, but we tried to keep it as simple as possible. Keep in
mind that mapbayr always expect a **pair of sigma values** for each type
of dependent variable: the **first** value for proportional error, the
**second** for additive.

Two situations can be distinguished:

1.  You only have one type of concentration to fit, and you did not use
    the \[OBS\] assignment in `$CMT`.

Simply write **one pair** of sigma values to describe proportional and
additive error on your concentrations. This error model will be
automatically applied to the compartment where observations were
recorded in your dataset (i.e. value of CMT when MDV = 0).

``` c
$SIGMA 0.111 0 // proportional error 
```

``` c
$SIGMA 0 0.222 // (log) additive error
```

``` c
$SIGMA 0.333 0.444 // mixed error
```

2.  You have multiple types of DV (parent and metabolite), and/or you
    used the \[OBS\] assignment in `$CMT`.

Write as many **pairs of sigma values** as there are compartments
assigned with \[OBS\] in `$CMT`. The order of the pair must respect the
order in which compartments were assigned. To put it more clearly, the
sigma matrix will be interpreted as such whatever the model :

| N° in the SIGMA matrix diagonal |                      Associated error                      |
|:-------------------------------:|:----------------------------------------------------------:|
|                1                | Proportional on concentrations in the 1st cmt with \[OBS\] |
|                2                |   Additive on concentrations in the 1st cmt with \[OBS\]   |
|                3                | Proportional on concentrations in the 2nd cmt with \[OBS\] |
|                4                |   Additive on concentrations in the 2nd cmt with \[OBS\]   |

``` c
//example: correlated proportional error between parent and metabolite
$SIGMA @block
0.050 // proportional error on parent drug
0.000 0.000 // additive error on parent drug
0.100 0.000 0.200 // proportional error on metabolite
0.000 0.000 0.000 0.000 // additive error on metabolite
// reminder: sigma values can be recorded in multiple $SIGMA blocks
```

### 6. `$TABLE` block or `$ERROR` block

- Mandatory:
  - Refer the concentration variable to fit as `DV`. Mind the code,
    especially if concentrations are observed in multiple compartments.
  - Express log-additive error models as exponential. This way,
    concentrations will automatically be log-transformed during the
    optimization process, with no necessity to prior log-transform your
    concentration.

``` c
$TABLE
double DV  = (CENTRAL / VC) * exp(EPS(2)) ;
```

- For fitting parent drug and metabolite simultaneously, refer to them
  as PAR and MET, and define DV accordingly (only DV will be used during
  the optimization process, but PAR and MET variables are mandatory for
  post-processing internal functions)

``` c
$TABLE
double PAR = (CENT_PAR / V) * (1 + EPS(1)) ;
double MET = (CENT_MET / V) * (1 + EPS(3)) ;
double DV = PAR ;
if(self.cmt == 4) DV = MET ; 
// reminder: use "self.cmt" to internaly refer to a compartment in a mrgsolve model code. 
```

Note that mapbayr does not strictly rely on this `$ERROR` block to
define the residual error internally and compute the objective function

NEWS.md
# mapbayr 0.9.0

## New features

### Data Helpers
* New `adm_rows()` and `obs_rows()` replace and improve `adm_lines()` and `obs_lines()`, respectively. See `?data_helpers` for a comprehensive documentation (#175).  
* In `adm_rows()`, `obs_rows()`, and `add_covariates()`, the first argument `x`:  
  - can be missing, which enables the creation of a new dataset from scratch.
  - accepts a data.frame, which enables the modification of a pre-existing dataset (#155). For example: 
    ```R
    adm_rows(amt = 100, cmt = 1) %>% 
      obs_rows(time = 24, cmt = 2, DV = 0.123)
    ```
  - still accepts an 'mrgsolve' model, which enables the modification of a dataset stored in the model arguments.
* In `adm_rows()` and `obs_rows()`, new argument `.datehour` in order to compute `time` as function of date and hours provided as character. The value passed to `.datehour` is parsed with `parse_datehour()` into a date-time value ("POSIXct"). For example: 
  ```R
  obs_rows(.datehour = c("2023/02/01 12:00", "2023/02/01 12:34"), DV = c(0.123, 0.456), cmt = 1)
  ```
* Data helpers now always rearrange data to fulfill the NM-TRAN compatibility and readability: filling missing covariate values with the last observation carried forward rule, relocation of NM-TRAN variables in the first positions etc...

### Below Limit of Quantification
* Data below the limit of quantification can now be handled with the so-called "M3 method" which consists in computing the likelihood of being below the limit of quantification. This is achieved when the variables `LLOQ` (lower limit of quantification, e.g. 0.22 mg/L) and `BLQ` (below limit of quantification, e.g. 1 or 0) are in the data (@pchelle, #182).
* This can be achieved by:  
  - adding the variables `LLOQ` and `BLQ` to the data by yourself.
  - adding the variable `LLOQ` to the data by yourself: `BLQ` will automatically be inferred from `LLOQ` and `DV`.
  - using `mapbayest(lloq = )` to automatically add the `LLOQ` and `BLQ` variables in the data.
* In `mapbayest()`, new argument `lloq` in order to add a variable `LLOQ` to the data. For example: `mapbayest(model, data, lloq = 0.22)`.

### Estimation features
* In `mapbayest()`, new argument `select_eta` in order to select the numbers of the ETAs to estimate. Default are ETAs related to an OMEGA not equal to zero. Non-selected ETAs will not be estimated and returned equal to zero. This can be useful in order to ignore the estimation of ETAs not of interest, e.g. in case of inter-occasion variability or non-identifiability. For example: `mapbayest(model, data, select_eta = c(1,3))` (#170).

* In `mapbayest()`, new argument `lambda` in order to modify the weight of the priors in the Bayesian estimation. This could be useful in order to flatten the priors with the objective to favor observed data instead of *a priori* information. For example: `mapbayest(model, data, lambda = 0.1)` to decrease the weight of priors of a ten-fold (#174).

## Minor changes and bug fixes
* Export `adm_rows.data.frame()`, `adm_rows.missing()`, `adm_rows.mrgmod()`, `obs_rows.data.frame()`, `obs_rows.missing()`, `obs_rows.mrgmod()`, and `add_covariates.data.frame()`, as new methods for data helpers. 
* Export `parse_datehour()`, used to parse arguments passed to `.datehour` in `adm_rows()` and `obs_rows()`.
* Export `filter.mrgmod()`, a method, wrapper around `dplyr::filter()` for dataset stored in mrgsolve model object ('mrgmod').
* Deprecate `adm_lines()` and `obs_lines()`. Stop exporting `adm_lines.mrgmod()` and `obs_lines.mrgmod()`.
* Stop exporting `see_data()`. Was deprecated since 0.4. Use `get_data()` instead.
* Suggests `lubridate`.
* In the final estimation object, new `arg.optim$select_eta` element, `arg.ofv.fix$omega_inv` now has the dimensions of the number of ETAs selected.
* OMEGA values equal to zero are allowed in the model and will be ignored during the estimations steps thanks to the new `select_eta` argument. This condition is not tested anymore with `check_mapbayr_model()`.
* `print.mapbayests()`, now only shows the estimated ETAs.
* In `hist()`, new argument `select_eta` in order to select the ETAs to plot. Default are ETAs estimated with `mapbayest()` (#167).
* `eta()` properly sorts vectors of length superior to 9 (#159).
* `use_posterior()` works if covariates had not been defined in data (#160).
* In `hist()`, `get_phi()` and `plot_phi()`, ETAs are now properly re-ordered if they are more than 9 (#165).
* Classification of absolute difference equal to zero now works (#166).
* Observations at time = 0 are now allowed (#168).
* Fix deprecations related to `tidyverse` packages (#171).
* Depends on `ggplot2 >= 3.4.0`.
* `add_covariates()` accepts empty arguments. 
* In `mapbayest()`, new argument `...` in order to fix compatibility issues, not used yet.

# mapbayr 0.8.0

## New features
* New `summarise_phi()` and `bar_phi()` summarizes the comparison of estimation of 'mapbayr' and 'NONMEM' (i.e. classifies it as Excellent/Acceptable/Discordant) and graphically represents it as a bar plot.
* New `eta()` generates numerical values named `ETA1, ETA2, ETA3...`, either from scratch, from a pre-existing vector or from a 'mrgsolve' model object.
* In `plot()`, `PREDICTION = c("IPRED", "PRED")` controls to plot either "PRED", "IPRED" or both (#113).
* In `add_covariates()`, `covariates` is relocated in last position, in the favor of `...` which now accepts covariate values. Calling `add_covariates(list(BW = 90))` will still works (with a warning) for the sake of compatibility but will be deprecated. Instead, just use `add_covariates(BW = 90)` or explicitly call `add_covariates(covariates = list(BW = 90))` if you want to pass covariate values as a list (#156).
* In `get_eta()`, `output = "num"` returns a matrix if multiple IDs are available instead of an error message (#145).

## Minor changes

R/mapbayests.R
#S3 Methods for `mapbayests` objects.

#' Print a mapbayests object
#'
#' @param x A \code{mapbayests} object.
#' @param ... additional arguments
#' @return print the results of the estimation to the console, and returns it invisibly.
#' @method print mapbayests
#' @export
print.mapbayests <- function(x, ...){
  NAME <- x$model@model
  nID <- length(x$arg.ofv.id)
  nOBS <- x$arg.ofv.id %>% map("idDV") %>% unname() %>% simplify() %>% length()
  nETA <- eta_length(x$model)
  ETA <- x$final_eta %>%
    bind_rows(.id = "ID") %>%
    as.data.frame() %>%
    utils::head()
  TAB <- utils::head(as.data.frame(x$mapbay_tab))

  cat("Model: ", NAME, "\n")
  cat("ID :", nID, " individual(s).\n")
  cat("OBS:", nOBS, " observation(s).\n")
  cat("ETA:", nETA, " parameter(s) to estimate.\n\n")
  cat("Estimates: \n")
  print(ETA)
  cat("\nOutput (", nrow(x$mapbay_tab) , " lines): \n", sep = "")
  print(TAB)
}


#' Return the mapbay_tab as a data.frame
#'
#' @param x A \code{mapbayests} object.
#' @param row.names,optional,... passed to as.data.frame
#'
#' @method as.data.frame mapbayests
#' @return a data.frame (the mapbay_tab from estimation)
#' @export
as.data.frame.mapbayests <- function(x, row.names = NULL, optional = FALSE, ...){
  as.data.frame(x$mapbay_tab, ...)
}


#' Plot predictions from mapbayests object
#'
#' @param x A \code{mapbayests} object.
#' @param ... additional arguments (passed to \code{\link{augment.mapbayests}})
#' @param PREDICTION plot either "IPRED", "PRED" or both.
#'
#' @return a `ggplot` object.
#'
#' @details
#' Use this function to plot the results of the estimations, in the form of concentration vs time profiles for every patient of the data set.
#' For additional modifications, you can:
#'  - see \code{\link{augment.mapbayests}} to modify the simulation output.
#'  - add extra `+function(...)` in order to modify the plot as a regular `ggplot2` object.
#'
#' @examples
#' est <- mapbayest(exmodel(ID = 1))
#' plot(est, end = 48) +
#'   ggplot2::labs(title = "Awesome prediction")
#'
#'
#' @method plot mapbayests
#' @export
plot.mapbayests <- function(x, ..., PREDICTION = c("IPRED", "PRED")){
  #  if(!inherits(x, "mapbayests")) stop("Provided object is not a mapbayests class object")

  if(is.null(x$aug_tab)){
    #  message("$aug_tab automatically provided. Consider executing augment() manually to save computational time or access options.")
    x <- augment(x, ...)
  }

  theme_custom <- function(...) {
    theme_bw(...) %+replace%
      theme(legend.position = "bottom",
            strip.background = element_rect(fill="white")
      )
  }

  predictions <- x$aug_tab %>%
    filter(.data$type %in% PREDICTION) %>%
    mutate(PREDICTION  = .data$type)

  gg <- predictions %>%
    ggplot(aes(.data$time, .data$value)) +
    geom_line(aes(col = .data$PREDICTION, linetype = .data$PREDICTION)) +
    theme_custom()+
    scale_color_manual(values= c(IPRED = "black", PRED = "deepskyblue1")) +
    scale_linetype_manual(values= c(IPRED = 1, PRED = 2))

  if(!is.null(predictions[["value_low"]]) & !is.null(predictions[["value_up"]])){
    data_ribbon <- predictions %>%
      filter(!(is.na(.data$value_low) & is.na(.data$value_up)))

    gg <- gg +
      geom_ribbon(aes(ymin = .data$value_low, ymax = .data$value_up, fill = .data$PREDICTION), data = data_ribbon, alpha = 0.3) +
      scale_fill_manual(values= c(IPRED = "black", PRED = "deepskyblue1"))
  }

  observations <- x$mapbay_tab %>%
    filter(.data$evid==0, !(.data$mdv==1 & is.na(.data$DV))) %>%
    mutate(MDV = as.factor(.data$mdv))

  #MDV
  if(any(observations$mdv == 1)){
    gg <- gg+
      geom_point(data = observations, aes(y = .data$DV, shape = .data$MDV), fill = "black", size = 3)+
      scale_shape_manual(values= c(`0` = 21, `1` = 1))
  } else {
    gg <- gg+
      geom_point(data = observations, aes(y = .data$DV), fill = "black", size = 3, pch = 21)
  }

  #Facetting

  one_cmt <- length(obs_cmt(x$model)) == 1
  one_ID <- length(x$arg.ofv.id) == 1

  if(all(!one_cmt, !one_ID)) {
    gg <- gg+
      facet_grid(ID~cmt, scales = "free", labeller = label_both)
  }

  if(all(one_cmt, !one_ID)) {
    gg <- gg+
      facet_grid(ID~., scales = "free", labeller = label_both)
  }

  if(all(!one_cmt, one_ID)) {
    gg <- gg+
      facet_grid(.~cmt, scales = "free", labeller = label_both)
  }

  return(gg)

}

#' Plot posterior distribution of bayesian estimates
#'
#' @param x A \code{mapbayests} object.
#' @param ... additional arguments (not used)
#' @return a `ggplot` object.
#'
#' @details
#' Use this function to plot the results of the estimations, in the form of histograms with the *a priori* distribution in the background. For every parameter, the inter-individual variability is displayed, as well as the percentile of the patient in the corresponding distribution (if n = 1 patient).
#' For additional modifications, you can add extra `+function(...)` in order to modify the plot as a regular `ggplot2` object.
#'
#' @examples
#' est <- mapbayest(exmodel(ID = 1))
#' hist(est) +
#'   ggplot2::labs(title = "Awesome estimations")
#' @method hist mapbayests
#' @export
hist.mapbayests <- function(x, ...){

  # --- Eta tab
  eta_tab <- x$final_eta %>%
    bind_rows() %>%
    pivot_longer(everything())

  # --- Arguments tab
  arg_tab <- data.frame(
    om = odiag(x$model),
    name = eta_names(x$model),
    descr = eta_descr(x$model),
    lower = x$arg.optim$lower,
    upper = x$arg.optim$upper
  )

  # --- Density tab
  minlow <- min(arg_tab$lower)
  maxup <- max(arg_tab$upper)
  xvalues <- seq(minlow - 0.01, maxup + 0.01, 0.01)

  density_tab <- arg_tab %>%
    select(.name = .data$name,.om = .data$om) %>%
    pmap_dfr(function(.name, .om){
      data.frame(name = .name,
                 x = xvalues,
                 value = stats::dnorm(xvalues, mean = 0, sd = sqrt(.om)))
    })

  # --- Labels
  eta_labs <- paste0(arg_tab$descr,
                     "\nIIV = ", my_percent(sqrt(arg_tab$om)))
  # --- one ID
  if(length(x$final_eta) == 1){
    percentile <- map2_dbl(x$final_eta[[1]], sqrt(arg_tab$om), stats::pnorm, mean = 0)
    eta_labs <- paste0(eta_labs,
                       "\nID percentile = ", my_percent(percentile))
  }

  names(eta_labs) <- arg_tab$name

  ggplot() +
    facet_wrap("name", labeller = labeller(name = eta_labs)) +
    geom_area(aes(x = .data$x, y = .data$value), data = density_tab, fill = "skyblue", alpha = .3) +
    geom_line(aes(x = .data$x, y = .data$value), data = density_tab) +
    geom_segment(aes(x = .data$lower, xend = .data$lower), y = -0.03, yend = .1, data = arg_tab, linetype = 1, size = 1) +
    geom_segment(aes(x = .data$upper, xend = .data$upper), y = -0.03, yend = .1, data = arg_tab, linetype = 1, size = 1) +
    theme_bw() +
    theme(strip.background = element_rect(fill = "white"))+
    scale_y_continuous(name = NULL, breaks = NULL, labels = NULL)+
    scale_x_continuous(name = NULL, n.breaks = 10)+
    coord_cartesian(ylim = c(NA, max(density_tab$value)))+
    geom_rug(aes(x = .data$value), data = eta_tab)+
    geom_histogram(aes(x = .data$value, y = .data$..density..), data = eta_tab, alpha = .8, col = 'black', bins = 50)
}


#' Compute full PK profile prediction from mapbayr estimates.
#' @param x object to augment
#' @param ... additional arguments
#' @export
#' @return an augmented object (depending on the object passed).
augment <- function (x, ...) UseMethod("augment")

#' Compute full PK profile prediction from mapbayr estimates.
#'
#' @param x A \code{mapbayests} object.
#' @param data dataset to pass to mrgsolve for simulation (default is dataset used for estimation)
#' @param start,end,delta start, end and delta of simulation time passed to `mrgsim()` (see details)
#' @param ci a logical. If TRUE, compute a confidence interval around the prediction (default is FALSE)
#' @param ci_width a number between 0 and 100, width of the confidence interval (default is "90" for a 90%CI)
#' @param ci_method method to compute the confidence interval. Can be "delta" (the default) to use the Delta approximation. Alternatively "simulations" for a more accurate approach, but also more time-consuming.
#' @param ci_sims number of replicates to simulate in order to derive the confidence interval (default is 500)
#' @param ... additional arguments passed to `mrgsim()`
#'
#' @method augment mapbayests
#' @return a `mapbayests` object, augmented of an `aug_tab` data.frame.
#' @details
#' This function is called in the background by `plot()` in order to simulate the full PK profile, and return a `mapbayests` object with an additional `aug_tab` data.frame inside. The latter is used with by the plot method.
#' The time grid, for each PK profile (i.e. patient) is defaulted with the minimum time in the dataset for `start` and the maximum time in the dataset +20% for `end`. `delta` is a power of 10 (e.g. 0.1, 1, 10 etc...), automatically chosen to render visually appealing graphs with a reasonable computing time (about 200 time points).
#' Additional arguments can be passed to `mrgsim()` through `...`. Note that `recsort` is set to 3 (see mrgsolve documentation for more details).
#'
#' @examples
#' #x is the result of `mapbayest()`.
#' #Default plot is returned by:
#' # plot(x)
#' #Argument passed to `plot()` are passed to `augment()` in the background:
#' # plot(x, end = 240, ci = TRUE)
#' #Save the augmented object if simulation time is long
#' # x2 <- augment(x, ci = TRUE, ci_method = "simulations", ci_sims = 10000) %>%
#' # plot(x2)
#'
#' @export
augment.mapbayests <- function(x, data = NULL, start = NULL, end = NULL, delta = NULL, ci = FALSE, ci_width = 90, ci_method = "delta", ci_sims = 500, ...){
  if(is.null(data)){
    idata <- get_data.mapbayests(x, output = "list")
  } else {
    idata <- data %>%
      check_mapbayr_data() %>%
      split_mapbayr_data()
  }

  if(is.null(start)){
    start <- unname(sapply(idata, function(x) min(x$time)))
    #A vector. For each ID, possibly a different start time.
  }

  if(is.null(end)){
    end <- unname(sapply(idata, function(x) max(x$time)))
    #A vector. For each ID, possibly a different end time.
    end <- end * 1.2
    #By default, +20% of last obs or dosing.
  }

  if(is.null(delta)){
    .delta <- (end - start)/200 #approximately 200 points per graph
    delta <- 10^(round(log10(abs(.delta)))) #rounded to the closer 10 (0.1, 1, 10 etc...)
    #A vector. For each ID, possibly a different delta.
  }

  fitcmt <- fit_cmt(x$model, get_data(x))

  if(length(fitcmt) > 1){
    toreq <- "PAR,MET"
  } else {
    toreq <- "DV"
  }

  tocarry <- c("evid")

  do_sims <- function(mods, data = idata, ...){
    sims <- list(
      x = mods,
      data = data, start = start, end = end, delta = delta
    ) %>%
      pmap(mrgsim_df, carry_out = tocarry, Request = toreq, recsort = 3, obsaug = TRUE, ... = ...)

    map(sims, ~ .x %>%
      filter(.data$evid %in% c(0,2)) %>%
      select(-any_of("cmt")) %>% #should not be carried normally but who knows...
      pivot_longer(any_of(c("DV", "PAR", "MET"))) %>%
      mutate(cmt = ifelse(.data$name %in% c("DV", "PAR"), fitcmt[1], fitcmt[2])) %>%
      arrange(.data$ID, .data$time, .data$cmt)
      )
  }

  do_augment <- function(x, type, ...){
    stopifnot(type %in% c("ipred", "pred"))

    mods <- switch(type,
                   "ipred" = use_posterior(x, update_eta = TRUE, update_omega = TRUE, update_cov = FALSE, simplify = FALSE),
                   "pred" = use_posterior(x, update_eta = FALSE, update_omega = FALSE, update_cov = FALSE, simplify = FALSE, .zero_re = "sigma"))

    initpreds <- do_sims(map(mods, zero_re), ... = ...)

    if(ci){
      if(ci_method == "delta"){
        etanames <- eta_names(mods[[1]])
        init_etas <- map(mods, ~unlist(param(.x)[etanames]))
        dsteps <- log(1+1e-8) #directly add 1e-8 because working on eta so will return into a multiplicative effect on parameter scale
        new_etas <- map2(init_etas, dsteps, ~.x+.y)
        new_models_etas <- map2(mods, new_etas, function(M,E){
          map(seq_along(E), ~zero_re(param(M, E[.x]))) %>% set_names(etanames)
        }) %>% transpose()
        new_preds <- map(new_models_etas, do_sims,  ... = ...) %>% transpose() #1 item = 1 indiv
        jacobians <- list(new_preds, dsteps, initpreds) %>% #1 item = 1 indiv
          pmap(function(new, step, ini){
            map2_dfc(new, step, ~(.x[["value"]]-ini[["value"]])/.y) %>% as.matrix()
          })
        varcovs <- map(mods, omat, make = TRUE) #IIV or uncertainty, depending on the update
        errors <- map2(jacobians, varcovs, ~ znorm(ci_width) * sqrt(diag(.x %*% .y %*% t(.x))))
        initpreds <- map2(initpreds, errors, ~mutate(.x,
                                                     value_low = .data[["value"]] - .y,
                                                     value_up = .data[["value"]] + .y))
      }

      if(ci_method == "simulations"){
        new_idatas <- map(idata, data_nid, n = ci_sims)
        new_sims <- do_sims(mods, data = new_idatas, ... = ...)
        LOW <- map(new_sims, ~ .x %>% prepare_summarise() %>% summarise(v = quantile(.data$value, ci2q(ci_width))) %>% pull("v"))
        UP <- map(new_sims, ~ .x %>% prepare_summarise() %>% summarise(v = quantile(.data$value, 1-ci2q(ci_width))) %>% pull("v"))
        initpreds <- pmap(list(initpreds, LOW, UP), function(ini, low, up){
          mutate(ini, value_low = low, value_up = up)
        })
      }
    }
    initpreds
  }

  ipred <- do_augment(x, type = "ipred", ... = ...) %>% bind_rows()
  pred  <- do_augment(x, type = "pred", ... = ...) %>% bind_rows()

  x$aug_tab <- bind_rows(list(IPRED = ipred, PRED = pred), .id = "type") %>%
    as_tibble() %>%
    arrange(.data$ID, .data$time, .data$cmt, .data$type)

  class(x) <- "mapbayests"
  return(x)
}

data_nid <- function(data, n){
  bind_rows(lapply(seq_len(n), function(x) mutate(data, ID = x)))
}

prepare_summarise <- function(data){
  data %>%
    group_by(.data$ID) %>%
    mutate(rowID = rank(.data$ID, ties.method = "first")) %>%
    group_by(.data$rowID)
}
